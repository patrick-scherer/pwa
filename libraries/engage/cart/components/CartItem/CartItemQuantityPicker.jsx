// @flow
import * as React from 'react';
import { i18n, isNumeric } from '@shopgate/engage/core';
import { inputStyle } from './CartItemQuantityPicker.style';

type Props = {
  editMode?: boolean,
  onChange?: (quantity: number) => void,
  onToggleEditMode?: (editMode: boolean) => void,
  quantity?: number,
}

type State = {
  quantity: number | string,
}

/**
 * The Quantity Picker component.
 */
export class CartItemQuantityPicker extends React.Component<Props, State> {
  static defaultProps = {
    editMode: false,
    onChange: () => { },
    quantity: 1,
    onToggleEditMode: () => { },
  };

  /**
   * Constructor.
   * @param {Object} props The component props.
   */
  constructor(props: Props) {
    super(props);

    this.regex = /^([0-9]{0,4})$/;

    this.state = {
      quantity: this.initialQuantity,
    };
  }

  /**
   * Called after mount. Focuses the input if the edit mode is active.
   */
  componentDidMount() {
    if (this.props.editMode && this.input) {
      this.input.focus();
    }

    if (this.input) {
      /**
       * Prevent the opening of the context menu when this
       * input is focused and the value is selected.
       */
      this.input.addEventListener('contextmenu', (event: MouseEvent) => {
        event.preventDefault();
        event.stopPropagation();
        return false;
      });
    }
  }

  /**
   * The componentWillReceiveProps lifecycle hook. I will bring the input into the correct state.
   * @param {Object} nextProps The next set of props.
   */
  UNSAFE_componentWillReceiveProps(nextProps: Props) {
    if (this.input) {
      if (nextProps.editMode) {
        this.input.focus();
      } else {
        this.input.blur();
      }
    }

    if (nextProps.quantity && this.props.quantity !== nextProps.quantity) {
      this.updateQuantityInState(nextProps.quantity);
    }
  }

  /**
   * Only update when certain state changes are made.
   * @param {Object} nextProps The next set of props.
   * @param {Object} nextState The next component state.
   * @returns {boolean}
   */
  shouldComponentUpdate(nextProps: Props, nextState: State) {
    return this.state.quantity !== nextState.quantity;
  }

  regex: RegExp;

  input: HTMLInputElement | null;

  /**
   * The default quantity.
   * @type {number}
   */
  defaultQuantity = 1;

  /**
   * Returns the initial quantity based on the props.
   * If the props are set to 0 or lower, it will fall back to 1.
   * @return {number}
   */
  get initialQuantity() {
    if (
      this.props.quantity
      && (this.props.quantity === 0 || this.props.quantity < this.defaultQuantity)
    ) {
      return this.defaultQuantity;
    }

    return this.props.quantity || this.defaultQuantity;
  }

  /**
   * Sets the component ref.
   * @param {HTMLElement} input The input field ref.
   */
  setRef = (input: HTMLInputElement | null) => {
    this.input = input;
  };

  /**
   * Event handler for the the onChange event of the input.
   * @param {Object} event The event object.
   */
  handleInputChange = (event: SyntheticInputEvent<HTMLInputElement>) => {
    const prevQuantity = this.state.quantity;
    let nextQuantity = event.target.value;

    if (this.regex.test(nextQuantity) === false) {
      // Reset the invalid value to the previous
      nextQuantity = prevQuantity;
    }

    // Update the quantity state with the new input.
    this.updateQuantityInState(nextQuantity);
  };

  /**
   * Event handler for the the onClick event of the input.
   * @param {Object} event The event object.
   */
  handleInputClick = (event: SyntheticInputEvent<HTMLInputElement>) => {
    // Prevent the native focus event ...
    event.stopPropagation();
    event.preventDefault();

    // ... and trigger it manually.
    // This way we avoid the select actions (copy, paste, cut) to become visible.
    if (this.input) {
      this.input.blur();
    }
    if (this.props.onToggleEditMode) {
      this.props.onToggleEditMode(true);
    }
  };

  /**
   * Event handler for the the onFocus event of the input.
   */
  handleInputFocus = () => {
    if (this.input) {
      this.input.select();
    }

    if (this.props.onToggleEditMode) {
      this.props.onToggleEditMode(true);
    }
  };

  /**
   * Event handler for the onSubmit event of the form which wraps the input.
   * @param {Object} event The event object.
   */
  handleSubmitForm = (event: SyntheticInputEvent<HTMLFormElement>) => {
    event.preventDefault();
    if (this.input) {
      this.input.blur();
    }
  };

  /**
   * Event handler for the the onBlur event of the input.
   * @param {Object} event The event object.
   */
  handleInputBlur = (event: SyntheticInputEvent<HTMLInputElement>) => {
    /**
     * Performs the actual logic to handle the blur event of the input.
     */
    const handleBlur = () => {
      const { onChange } = this.props;

      // Deactivate the edit mode
      if (this.props.onToggleEditMode) {
        this.props.onToggleEditMode(false);
      }

      if (this.props.quantity !== this.state.quantity) {
        if (onChange) {
          onChange(parseInt(this.state.quantity, 10));
        }
      }
    };

    if (!event.target.value) {
      // Set the quantity state back to default, if the value of the input is invalid.
      this.updateQuantityInState(this.defaultQuantity, handleBlur);
    } else {
      handleBlur();
    }
  };

  /**
   * Updates the quantity within the component state. It takes care,
   * that the state quantity is always a numeric value.
   * @param {string|number} quantity The new quantity
   * @param {Function} [callback] Callback for the setState call.
   */
  updateQuantityInState(quantity: number | string, callback?: () => void = () => { }) {
    const sanitizedQuantity = isNumeric(quantity) ? parseInt(quantity, 10) : '';

    this.setState({
      quantity: sanitizedQuantity,
    }, callback);
  }

  /**
   * Renders the component.
   * @return {JSX}
   */
  render() {
    return (
      <form onSubmit={this.handleSubmitForm}>
        <input
          ref={this.setRef}
          type="number"
          className={inputStyle}
          value={this.state.quantity}
          onChange={this.handleInputChange}
          onClick={this.handleInputClick}
          onFocus={this.handleInputFocus}
          onBlur={this.handleInputBlur}
          min={0}
          data-test-id="quantityPicker"
          aria-label={i18n.text('product.quantity')}
        />
      </form>
    );
  }
}
