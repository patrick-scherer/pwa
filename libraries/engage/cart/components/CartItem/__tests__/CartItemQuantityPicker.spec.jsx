import React from 'react';
import { mount } from 'enzyme';
import { CartItemQuantityPicker } from '../CartItemQuantityPicker';

describe('<CartItemQuantityPicker />', () => {
  it('should have an amount of 1 by default', () => {
    const wrapper = mount(<CartItemQuantityPicker />);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('input').prop('value')).toEqual('1');
  });

  it('should have an amount of 3 via prop', () => {
    const wrapper = mount(<CartItemQuantityPicker quantity={3} />);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('input').prop('value')).toEqual('3');
  });

  it('should have an mount of 0, if 0 is supplied', () => {
    const wrapper = mount(<CartItemQuantityPicker quantity={0} />);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('input').prop('value')).toEqual('0');
  });

  it.skip('should reset the amount to 1 if set to lower than 0', () => {
    const wrapper = mount(<CartItemQuantityPicker quantity={-1} />);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('input').prop('value')).toEqual(1);
  });

  it.skip('should fall back to the previous value when the last value was invalid', () => {
    const wrapper = mount(<CartItemQuantityPicker quantity={124} />);
    const input = wrapper.find('input');

    input.simulate('change', { target: { value: '124.5' } });

    expect(wrapper).toMatchSnapshot();
    expect(input.prop('value')).toBe(124);
  });

  describe('Given editMode prop is handled correctly', () => {
    let wrapper;
    let input;

    beforeEach(() => {
      wrapper = mount(<CartItemQuantityPicker />);

      input = wrapper.instance().input.current;
      jest.spyOn(input, 'focus');
      jest.spyOn(input, 'blur');
    });

    it('should focus the input, if the editMode is enabled via prop', () => {
      wrapper.setProps({
        editMode: true,
      });

      expect(wrapper).toMatchSnapshot();
      expect(input.focus).toHaveBeenCalled();
    });

    it('should blur the input, if the editMode is disabled via prop', () => {
      wrapper.setProps({
        editMode: false,
      });

      expect(wrapper).toMatchSnapshot();
      expect(input.blur).toHaveBeenCalled();
    });
  });

  describe('Given onChange callback is triggered correctly', () => {
    let wrapper;
    let input;
    let onChangeMock;

    beforeEach(() => {
      onChangeMock = jest.fn();

      wrapper = mount(<CartItemQuantityPicker quantity={2} onChange={onChangeMock} />);

      input = wrapper.find('input');
    });

    it('should trigger the callback when the input changed', () => {
      input.simulate('focus');
      input.simulate('change', { target: { value: '123' } });
      input.simulate('blur');

      expect(wrapper).toMatchSnapshot();
      expect(onChangeMock.mock.calls.length).toBe(1);
    });

    it('should not trigger the callback when the input didn\'t change', () => {
      input.simulate('focus');
      input.simulate('change', { target: { value: '2' } });
      input.simulate('blur');

      expect(wrapper).toMatchSnapshot();
      expect(onChangeMock.mock.calls.length).toBe(0);
    });
  });

  describe('Given onToggleEditMode callback is triggered correctly', () => {
    let wrapper;
    let input;
    let onToggleEditModeMock;

    beforeEach(() => {
      onToggleEditModeMock = jest.fn();

      wrapper = mount(<CartItemQuantityPicker onToggleEditMode={onToggleEditModeMock} />);

      input = wrapper.find('input');
    });

    it('should trigger the callback when the input was focused', () => {
      input.simulate('focus');

      expect(wrapper).toMatchSnapshot();
      expect(onToggleEditModeMock).lastCalledWith(true);
    });

    it('should trigger the callback when the input was blurred', () => {
      input.simulate('focus');
      input.simulate('blur');

      expect(wrapper).toMatchSnapshot();
      expect(onToggleEditModeMock).lastCalledWith(false);
    });
  });
});
