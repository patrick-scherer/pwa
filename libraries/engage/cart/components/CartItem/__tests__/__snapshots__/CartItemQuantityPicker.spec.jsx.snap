// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartItemQuantityPicker /> Given editMode prop is handled correctly should blur the input, if the editMode is disabled via prop 1`] = `
<CartItemQuantityPicker
  editMode={false}
  onChange={[Function]}
  onToggleEditMode={[Function]}
  quantity={1}
>
  <form
    onSubmit={[Function]}
  >
    <input
      aria-label="product.quantity"
      className={
        Object {
          "data-css-1bncmx2": "",
        }
      }
      data-test-id="quantityPicker"
      min={0}
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      type="number"
      value={1}
    />
  </form>
</CartItemQuantityPicker>
`;

exports[`<CartItemQuantityPicker /> Given editMode prop is handled correctly should focus the input, if the editMode is enabled via prop 1`] = `
<CartItemQuantityPicker
  editMode={true}
  onChange={[Function]}
  onToggleEditMode={[Function]}
  quantity={1}
>
  <form
    onSubmit={[Function]}
  >
    <input
      aria-label="product.quantity"
      className={
        Object {
          "data-css-1bncmx2": "",
        }
      }
      data-test-id="quantityPicker"
      min={0}
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      type="number"
      value={1}
    />
  </form>
</CartItemQuantityPicker>
`;

exports[`<CartItemQuantityPicker /> Given onChange callback is triggered correctly should not trigger the callback when the input didn't change 1`] = `
<CartItemQuantityPicker
  editMode={false}
  onChange={[MockFunction]}
  onToggleEditMode={[Function]}
  quantity={2}
>
  <form
    onSubmit={[Function]}
  >
    <input
      aria-label="product.quantity"
      className={
        Object {
          "data-css-1bncmx2": "",
        }
      }
      data-test-id="quantityPicker"
      min={0}
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      type="number"
      value={2}
    />
  </form>
</CartItemQuantityPicker>
`;

exports[`<CartItemQuantityPicker /> Given onChange callback is triggered correctly should trigger the callback when the input changed 1`] = `
<CartItemQuantityPicker
  editMode={false}
  onChange={
    [MockFunction] {
      "calls": Array [
        Array [
          123,
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  onToggleEditMode={[Function]}
  quantity={2}
>
  <form
    onSubmit={[Function]}
  >
    <input
      aria-label="product.quantity"
      className={
        Object {
          "data-css-1bncmx2": "",
        }
      }
      data-test-id="quantityPicker"
      min={0}
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      type="number"
      value={123}
    />
  </form>
</CartItemQuantityPicker>
`;

exports[`<CartItemQuantityPicker /> Given onToggleEditMode callback is triggered correctly should trigger the callback when the input was blurred 1`] = `
<CartItemQuantityPicker
  editMode={false}
  onChange={[Function]}
  onToggleEditMode={
    [MockFunction] {
      "calls": Array [
        Array [
          true,
        ],
        Array [
          false,
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  quantity={1}
>
  <form
    onSubmit={[Function]}
  >
    <input
      aria-label="product.quantity"
      className={
        Object {
          "data-css-1bncmx2": "",
        }
      }
      data-test-id="quantityPicker"
      min={0}
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      type="number"
      value={1}
    />
  </form>
</CartItemQuantityPicker>
`;

exports[`<CartItemQuantityPicker /> Given onToggleEditMode callback is triggered correctly should trigger the callback when the input was focused 1`] = `
<CartItemQuantityPicker
  editMode={false}
  onChange={[Function]}
  onToggleEditMode={
    [MockFunction] {
      "calls": Array [
        Array [
          true,
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  quantity={1}
>
  <form
    onSubmit={[Function]}
  >
    <input
      aria-label="product.quantity"
      className={
        Object {
          "data-css-1bncmx2": "",
        }
      }
      data-test-id="quantityPicker"
      min={0}
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      type="number"
      value={1}
    />
  </form>
</CartItemQuantityPicker>
`;

exports[`<CartItemQuantityPicker /> should fall back to its default of 1, if 0 is supplied 1`] = `
<CartItemQuantityPicker
  editMode={false}
  onChange={[Function]}
  onToggleEditMode={[Function]}
  quantity={0}
>
  <form
    onSubmit={[Function]}
  >
    <input
      aria-label="product.quantity"
      className={
        Object {
          "data-css-1bncmx2": "",
        }
      }
      data-test-id="quantityPicker"
      min={0}
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      type="number"
      value={1}
    />
  </form>
</CartItemQuantityPicker>
`;

exports[`<CartItemQuantityPicker /> should fall back to the previous value when the last value was invalid 1`] = `
<CartItemQuantityPicker
  editMode={false}
  onChange={[Function]}
  onToggleEditMode={[Function]}
  quantity={124}
>
  <form
    onSubmit={[Function]}
  >
    <input
      aria-label="product.quantity"
      className={
        Object {
          "data-css-1bncmx2": "",
        }
      }
      data-test-id="quantityPicker"
      min={0}
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      type="number"
      value={124}
    />
  </form>
</CartItemQuantityPicker>
`;

exports[`<CartItemQuantityPicker /> should have an amount of 1 by default 1`] = `
<CartItemQuantityPicker
  editMode={false}
  onChange={[Function]}
  onToggleEditMode={[Function]}
  quantity={1}
>
  <form
    onSubmit={[Function]}
  >
    <input
      aria-label="product.quantity"
      className={
        Object {
          "data-css-1bncmx2": "",
        }
      }
      data-test-id="quantityPicker"
      min={0}
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      type="number"
      value={1}
    />
  </form>
</CartItemQuantityPicker>
`;

exports[`<CartItemQuantityPicker /> should have an amount of 3 via prop 1`] = `
<CartItemQuantityPicker
  editMode={false}
  onChange={[Function]}
  onToggleEditMode={[Function]}
  quantity={3}
>
  <form
    onSubmit={[Function]}
  >
    <input
      aria-label="product.quantity"
      className={
        Object {
          "data-css-1bncmx2": "",
        }
      }
      data-test-id="quantityPicker"
      min={0}
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      type="number"
      value={3}
    />
  </form>
</CartItemQuantityPicker>
`;

exports[`<CartItemQuantityPicker /> should reset the amount to 1 if set to lower than 0 1`] = `
<CartItemQuantityPicker
  editMode={false}
  onChange={[Function]}
  onToggleEditMode={[Function]}
  quantity={-1}
>
  <form
    onSubmit={[Function]}
  >
    <input
      aria-label="product.quantity"
      className={
        Object {
          "data-css-1bncmx2": "",
        }
      }
      data-test-id="quantityPicker"
      min={0}
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      type="number"
      value={1}
    />
  </form>
</CartItemQuantityPicker>
`;
